### PR TITLE
feat: Add BaseDAO improvements and test reorganization

### PR DESCRIPTION
**Pull request by aminghadersohi**
### SUMMARY

This PR enhances the BaseDAO layer to support structured programmatic filtering, enabling external API integrations (like MCP services) to query Superset data without Flask-AppBuilder dependencies. The changes are fully backward compatible with zero breaking changes.

Next PR on top of this: https://github.com/apache/superset/pull/35163

#### Key Enhancements

**1. Structured Filtering System**
- Added `ColumnOperator` model with 14 filter operators (eq, neq, gt, gte, lt, lte, like, ilike, in, nin, null, nnull, sw, nsw)
- Enables type-safe, programmatic filtering without FAB web dependencies
- Supports complex queries with multiple filters and operators

**2. Enhanced Query Methods**
- New `list()` method with comprehensive filtering, sorting, pagination, and search
- New `count()` method for filtered record counting
- Schema discovery via `get_filterable_columns_and_operators()` for API documentation

**3. UUID Support & Flexible ID Handling**
- Automatic UUID type conversion for gradual migration to UUID primary keys
- Support for non-standard ID field names across models
- Enhanced `find_by_id()`/`find_by_ids()` with string/UUID flexibility

**4. Model-Specific Extensions**
- ChartDAO: Added filterable fields for `tags`, `owner`, `viz_type`, `datasource_name`
- DashboardDAO: Added filterable fields for `tags`, `owner`, `published`, `favorite`
- DatasetDAO: Added filterable fields for `tags`, `owner`

#### Design Rationale

The existing DAO layer lacked programmatic filtering capabilities needed by external services. This enhancement:
- Maintains complete backward compatibility (all existing code continues working)
- Follows Superset's patterns and conventions
- Provides type safety with MyPy compliance
- Enables self-documenting APIs through schema discovery
- Supports the ongoing UUID migration initiative

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A - Backend infrastructure change with no UI impact

### TESTING INSTRUCTIONS

1. **Run the comprehensive test suite:**
   ```bash
   # Run the new DAO tests (2400+ lines of test coverage)
   pytest tests/unit_tests/dao/
   ```

2. **Verify backward compatibility:**
   ```bash
   # Existing DAO tests should all pass unchanged
   pytest tests/unit_tests/daos/
   pytest tests/integration_tests/charts/
   pytest tests/integration_tests/dashboards/
   pytest tests/integration_tests/datasets/
   ```

3. **Test programmatic filtering (Python shell example):**
   ```python
   from superset.daos.chart import ChartDAO
   from superset.daos.base import ColumnOperator, ColumnOperatorEnum
   
   # Test filtering
   charts, total = ChartDAO.list(
       column_operators=[
           ColumnOperator(col="viz_type", opr=ColumnOperatorEnum.eq, value="line")
       ],
       page_size=10
   )
   
   # Test schema discovery
   schema = ChartDAO.get_filterable_columns_and_operators()
   assert "viz_type" in schema
   ```

4. **Verify type checking:**
   ```bash
   pre-commit run mypy
   ```

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://app.shortcut.com/preset/story/91603/add-basedao-improvements-to-reduce-reliance-on-fab
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
  - New DAO methods: `list()`, `count()`, `get_filterable_columns_and_operators()`
  - New filtering system via `ColumnOperator` and `ColumnOperatorEnum`
  - Fully backward compatible - no breaking changes
- [ ] Removes existing feature or API

**Additional Notes:**
- This PR is the foundation for the MCP (Model Context Protocol) service integration
- Comprehensive test coverage: 2400+ lines of tests covering all operators and edge cases
- Follows Superset's ongoing initiatives: UUID migration, type safety (MyPy), and API standardization
- Zero performance impact on existing code paths - new features are opt-in via optional parameters

### CODE REVIEW

**Comment by villebro** [comment]
This is great work and will also be very useful for list view filters.

File: superset/daos/base.py [hunk 1]
```diff
@@ -251,3 +411,247 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+
+    @classmethod
+    def get_filterable_columns_and_operators(cls) -> Dict[str, List[str]]:
+        """
+        Returns a dict mapping filterable columns (including hybrid/computed fields if
+        present) to their supported operators. Used by MCP tools to dynamically expose
+        filter options. Custom fields supported by the DAO but not present on the model
+        should be documented here.
+        """
+        from sqlalchemy.ext.hybrid import hybrid_property
```
Comment (mistercrunch) [inline superset/daos/base.py]: NIT: i think top-of-module imports are generally better, except for some exceptions, I keep catching claude doing this as I think it means fewer edits for it. [I think] exception are around potential circular deps or importing heavy modules

File: superset/daos/base.py [hunk 2]
```diff
@@ -251,3 +411,247 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+        hybrids = {
+            name: attr
+            for name, attr in vars(cls.model_cls).items()
+            if isinstance(attr, hybrid_property)
+        }
+        # You may add custom fields here, e.g.:
+        # custom_fields = {"tags": ["eq", "in_", "like"], ...}
+        custom_fields: Dict[str, List[str]] = {}
+        # Map SQLAlchemy types to supported operators
+        type_operator_map = {
```
Comment (mistercrunch) [inline superset/daos/base.py]: NIT: could be great as a module constant - or class-level prop if there's a need to override in subclasses, but i think it's static module constant probably best

File: superset/daos/chart.py [hunk 3]
```diff
@@ -39,6 +39,20 @@
 class ChartDAO(BaseDAO[Slice]):
     base_filter = ChartFilter
 
+    @classmethod
+    def get_filterable_columns_and_operators(cls) -> Dict[str, List[str]]:
+        filterable = super().get_filterable_columns_and_operators()
+        # Add custom fields for charts
+        filterable.update(
+            {
```
Comment (mistercrunch) [inline superset/daos/chart.py]: NIT: also could fit as a module constant, this way it's not regenerated at every call

File: tests/unit_tests/dao/base_dao_test.py [hunk 4]
```diff
@@ -37,6 +79,1924 @@ class TestDAOWithNoneModel(BaseDAO[MockModel]):
...
+def test_base_dao_edge_cases(app_context: None) -> None:
+    """Test BaseDAO edge cases and error conditions."""
+    # Test create without item or attributes
+    created = UserDAO.create()
+    assert created is not None
+    # User model has required fields, so we expect them to be None
+    assert created.username is None
+
+    # Don't commit - would fail due to constraints
+    db.session.rollback()
```
Comment (mistercrunch) [inline tests/unit_tests/dao/base_dao_test.py]: I think technically you might have integration tests in here (seem to use the database, didn't analyze to see if there are module-level mocks setup). I don't think we've been great at segmenting properly, so there are probably integration tests in unit tests in Superset and vice-versa. So in theory might belong in `tests/unit_tests/` (?)

**Review by mistercrunch — COMMENTED** [review]
Few small nits but overall solid code pushing the right logic where it belongs, excited to see base classes get strengthened up and more test coverage!

Wondering if this refactor means some code in the derived classes will automatically benefit or whether we could push subclasses to use the new methods(?)

File: superset/daos/base.py (hunk 1 above)
Comment (aminghadersohi) [reply superset/daos/base.py]: you're absolutely right. I

File: superset/daos/base.py (hunk 2 above)
Comment (aminghadersohi) [reply superset/daos/base.py]: you're right. better defined once at module load time  rather than being recreated every time the get_filterable_columns_and_operators method is called.

File: tests/unit_tests/dao/base_dao_test.py (hunk 4 above)
Comment (aminghadersohi) [reply tests/unit_tests/dao/base_dao_test.py]: yeah i think you're right. some of these are integration tests, and should be in `tests/integration_tests` as such. I will refactor.

Comment (aminghadersohi) [reply tests/unit_tests/dao/base_dao_test.py]: Crunching...

**Comment by aminghadersohi** [comment]
> Few small nits but overall solid code pushing the right logic where it belongs, excited to see base classes get strengthened up and more test coverage!
> 
> Wondering if this refactor means some code in the derived classes will automatically benefit or whether we could push subclasses to use the new methods(?)

Thanks! 

The added code does not affect existing usages of daos - so subclasses don't use this code yet. I think a future refactor would be beneficial.

File: requirements/development.in [hunk 5]
```diff
@@ -16,5 +16,5 @@
 # specific language governing permissions and limitations
 # under the License.
 #
--e .[development,bigquery,druid,duckdb,gevent,gsheets,mysql,postgres,presto,prophet,trino,thumbnails]
+-e .[development,bigquery,druid,duckdb,gevent,gsheets,mysql,postgres,presto,prophet,pydantic,trino,thumbnails]
```
[the remark this answers is not in the document]
Comment (mistercrunch) [reply requirements/development.in]: The bot is right here, `pydantic` is not required in this file since it's part of the core dependencies in `pyproject.toml`. The references here are pointing to a set of `project.optional-dependencies` (same as `extra_requires` in old setup.py). Can safely remove `pydantic` and re-run `./scripts/uv-pip-compile.sh` which should have no effect since pydantic is kind of a no-op here.

File: pyproject.toml [hunk 6]
```diff
@@ -84,6 +84,7 @@ dependencies = [
     "pgsanity",
     "Pillow>=11.0.0, <12",
     "polyline>=2.0.0, <3.0",
+    "pydantic>=2.8.0",
```
Comment (mistercrunch) [inline pyproject.toml:87]: this is sufficient here, meaning adding the dep here will affect both requirements/base.txt and requirements/development.txt automatically when re-running `./scripts/uv-pip-compile.sh`

File: superset/daos/base.py [hunk 7]
```diff
@@ -32,6 +51,100 @@
...
+            ),
+            cls.gt: lambda col, val: col > val,
+            cls.gte: lambda col, val: col >= val,
+            cls.lt: lambda col, val: col < val,
+            cls.lte: lambda col, val: col <= val,
+            cls.like: lambda col, val: col.like(f"%{val}%"),
+            cls.ilike: lambda col, val: col.ilike(f"%{val}%"),
+            cls.is_null: lambda col, _: col.is_(None),
+            cls.is_not_null: lambda col, _: col.isnot(None),
+        }
```
Comment (dpgaspar) [inline superset/daos/base.py]: WDYT about just making this a dict, would be slightly faster since we would avoid a function call

File: superset/daos/base.py [hunk 8]
```diff
@@ -83,39 +196,138 @@ def find_by_id_or_uuid(
...
+        id_col = getattr(cls.model_cls, column, None)
         if id_col is None or not model_ids:
             return []
+
+        # Convert IDs to appropriate types based on column type
+        converted_ids: list[str | int | uuid_lib.UUID] = []
+        for id_val in model_ids:
+            converted_value = cls._convert_value_for_column(id_col, id_val)
+            if converted_value is not None:
+                converted_ids.append(converted_value)
```
Comment (dpgaspar) [inline superset/daos/base.py:324]: nit: it seems we are assuming that all values are of the same type, should we make sure this is always true?

File: superset/daos/base.py [hunk 9]
```diff
@@ -251,3 +455,205 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+        for c in column_operators:
+            if not isinstance(c, ColumnOperator):
+                continue
+            col = c.col
+            opr = c.opr
+            value = c.value
+            if not col or not hasattr(cls.model_cls, col):
+                model_name = cls.model_cls.__name__ if cls.model_cls else "Unknown"
+                logging.error(
+                    f"Invalid filter: column '{col}' does not exist on {model_name}"
```
Comment (dpgaspar) [inline superset/daos/base.py]: nit: we should not use f strings when logging for performance reasons:

The f-string is evaluated immediately, even if the log level is higher (e.g., INFO or WARNING) and the DEBUG message will never be logged.
This means you pay the cost of:
Building the string
Evaluating any expressions inside {}
Even if the message is dropped due to the log level.

let's use:
``` python
logging.error("Invalid filter: column '%s' does not exist on %s", col, model_name)
```

File: superset/daos/base.py [hunk 10]
```diff
@@ -251,3 +455,205 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+                raise ValueError(
+                    f"Invalid filter: column '{col}' does not exist on {model_name}"
+                )
+            column = getattr(cls.model_cls, col)
+            try:
+                # Always use ColumnOperatorEnum's apply method
+                operator_enum = ColumnOperatorEnum(opr)
+                query = query.filter(operator_enum.apply(column, value))
+            except Exception as e:
+                logging.error(f"Error applying filter on column '{col}': {e}")
```
Comment (dpgaspar) [inline superset/daos/base.py]: nit: let's use lazy formatting

File: superset/daos/base.py [hunk 11]
```diff
@@ -251,3 +455,205 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+            columns = []
+        for name in columns:
+            attr = getattr(cls.model_cls, name, None)
+            if attr is None:
+                continue
+            prop = getattr(attr, "property", None)
+            if isinstance(prop, ColumnProperty):
+                column_attrs.append(attr)
+            elif isinstance(prop, RelationshipProperty):
+                relationship_loads.append(joinedload(attr))
```
Comment (dpgaspar) [inline superset/daos/base.py:624]: It seems we will support many-to-many relations here, double check if pagination will work as expected on these cases.

do we have tests for relations?

File: tests/integration_tests/dao/base_dao_test.py [hunk 12]
```diff
@@ -0,0 +1,1397 @@
...
+        email="detached@example.com",
+        active=True,
+    )
+    db.session.add(user)
+    db.session.commit()
+
+    user_id = user.id
+
+    # Expunge to detach from session
+    db.session.expunge(user)
```
Comment (dpgaspar) [inline tests/integration_tests/dao/base_dao_test.py:490]: interesting test!

**Review by dpgaspar — COMMENTED** [review]
Cool!

File: requirements/development.in (hunk 5 above)
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply requirements/development.in]: ah yeah that makes sense. pydantic gets pulled as a part of fastmcp. This PR doesn't have fastMCP dep requirement but it does add pydantic usage to base dao. It wasn't really possible to keep using marshmallow because the mcp service uses the pydantic models directly and/or extends them.

[the remark this answers is not in the document]
Comment (mistercrunch) [reply requirements/development.in]: it doesn't need to be referenced in this particular file as it's referenced in pyproject.toml core dependencies.

File: superset/daos/base.py (hunk 8 above)
Comment (aminghadersohi) [reply superset/daos/base.py:324]: Good catch! You're absolutely right - we should validate type consistency. I'll add a check to ensure all converted IDs are of the same type, with a warning logged for mixed types. This will help catch bugs early while maintaining backward compatibility.

File: superset/daos/base.py [hunk 13]
```diff
@@ -83,39 +196,138 @@ def find_by_id_or_uuid(
...
+
+        try:
+            results = query.all()
+        except SQLAlchemyError as ex:
+            model_name = cls.model_cls.__name__ if cls.model_cls else "Unknown"
+            raise DAOFindFailedError(
+                f"Failed to find {model_name} with ids: {model_ids}"
+            ) from ex
+
+        return results
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/daos/base.py]: this was my mistake.

File: superset/daos/base.py (hunk 9 above)
Comment (aminghadersohi) [reply superset/daos/base.py]: good catch. i will fix.

File: superset/daos/base.py (hunk 11 above)
Comment (aminghadersohi) [reply superset/daos/base.py:624]: i had some initial trouble with using relations, so they aren't used right now by the mcp service in its discovery tools. but thats a good flag to revisit. i will write back my findings.

Comment (aminghadersohi) [reply superset/daos/base.py:624]: Good catch! You're right - `joinedload` with many-to-many relationships
  would break pagination. I've fixed it by counting before adding the
  relationship joins. This ensures:
  - Accurate counts (not inflated by joins)
  - Correct pagination (LIMIT/OFFSET on entities, not joined rows)

  We don't have tests for relationship loading in the list method - that
  should be added to prevent regression.

File: tests/integration_tests/dao/base_dao_test.py (hunk 12 above)
Comment (aminghadersohi) [reply tests/integration_tests/dao/base_dao_test.py:490]: if i recall correctly, i ran into some issues with detached objects while working on the mcp service and got claude code to write this test.

Comment (aminghadersohi) [reply tests/integration_tests/dao/base_dao_test.py:490]: 1. Real-world scenario: In web applications, objects often become
  detached between requests. For example:
    - Object retrieved in one request, modified in another
    - Objects stored in cache/session and later updated
    - Background tasks that operate on objects outside the request
  context
  2. SQLAlchemy complexity: Detached objects are a common source of bugs
  with SQLAlchemy. Without proper handling, you get errors like:
    - DetachedInstanceError
    - InvalidRequestError: Object is not bound to a Session
  3. API contract: This test documents that BaseDAO.update() explicitly
  supports detached objects. Removing the test would mean we're no longer
   guaranteeing this behavior, which could break existing code that
  relies on it.
  4. Defensive programming: It ensures the DAO layer is robust

File: tests/integration_tests/dao/base_dao_test.py [hunk 14]
```diff
@@ -0,0 +1,1397 @@
...
+
+    dashboard_ids = [dashboard.id for dashboard in dashboards]
+
+    DashboardDAO.delete(dashboards)
+    db.session.commit()
+
+    # Verify they're all gone
+    for dashboard_id in dashboard_ids:
+        found = DashboardDAO.find_by_id(dashboard_id, skip_base_filter=True)
+        assert found is None
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply tests/integration_tests/dao/base_dao_test.py:549]: this inside the loop.

File: tests/unit_tests/dao/base_dao_test.py [hunk 15]
```diff
@@ -37,6 +44,103 @@ class TestDAOWithNoneModel(BaseDAO[MockModel]):
...
+        elif ">=" in expected_sql_fragment:
+            assert ">=" in sql_str
+        elif "<" in expected_sql_fragment:
+            assert "<" in sql_str
+        elif "<=" in expected_sql_fragment:
+            assert "<=" in sql_str
+        elif "IS NULL" in expected_sql_fragment:
+            assert "IS NULL" in sql_str
+        elif "IS NOT NULL" in expected_sql_fragment:
+            assert "IS NOT NULL" in sql_str
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply tests/unit_tests/dao/base_dao_test.py]: You're right - these assertions are way too weak. They just check if
  certain characters exist anywhere in the SQL string, which could lead
  to false positives. Let me rewrite this test to properly validate the
  SQL generation

File: superset/daos/base.py [hunk 16]
```diff
@@ -83,52 +197,145 @@ def find_by_id_or_uuid(
...
+
+        # Validate type consistency for better error handling
+        if len(converted_ids) > 1:
+            types_found = set(map(type, converted_ids))
+            if len(types_found) > 1:
+                logger.warning(
+                    "Mixed ID types detected for %s: %s",
+                    cls.model_cls.__name__ if cls.model_cls else "Unknown",
+                    [t.__name__ for t in types_found],
+                )
```
Comment (aminghadersohi) [inline superset/daos/base.py]: @dpgaspar what do you think of this approach?

File: superset/daos/base.py [hunk 17]
```diff
@@ -251,3 +450,213 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+            for filter_class in custom_filters.values():
+                query = filter_class.apply(query, None)
+        if column_operators:
+            query = cls.apply_column_operators(query, column_operators)
+
+        # Count before adding relationship joins to avoid inflated counts
+        # with one-to-many or many-to-many relationships
+        total_count = query.count()
+
+        # Add relationship joins after counting
```
Comment (aminghadersohi) [inline superset/daos/base.py:656]: @dpgaspar moved the relationship joins to after the count

**Comment by aminghadersohi** [comment]
I have addressed all comments so far. Thank you for the very thorough review @dpgaspar !

File: superset/daos/base.py (hunk 9 above)
Comment (mistercrunch) [reply superset/daos/base.py]: oh right, I feel linters used to catch these ... maybe we lost that rule when moving from pylint to ruff? 

```
[tool.ruff]
select = ["E", "F", "G"]  # add "G" to enable flake8-logging-format rules
```

Comment (mistercrunch) [reply superset/daos/base.py]: I'll put my best bot on it!

Comment (mistercrunch) [reply superset/daos/base.py]: Oops:
```
Found 190 errors.
[*] 1 fixable with the `--fix` option.
```
alright, I just put 8 subagents on the task :)

Comment (mistercrunch) [reply superset/daos/base.py]: https://github.com/apache/superset/pull/35081

File: tests/integration_tests/dao/base_dao_test.py (hunk 12 above)
Comment (dpgaspar) [reply tests/integration_tests/dao/base_dao_test.py:490]: I like it!!

**Comment by aminghadersohi** [comment]
fyi i have rebased this and pushed up to help w upcoming stacked PRs

File: superset/daos/base.py (hunk 16 above)
Comment (dpgaspar) [reply superset/daos/base.py]: Looks good, but should it fail with ValueError?

File: superset/daos/base.py [hunk 18]
```diff
@@ -251,3 +457,231 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+        non-existent column.
+        """
+        if not column_operators:
+            return query
+        for c in column_operators:
+            if not isinstance(c, ColumnOperator):
+                continue
+            col = c.col
+            opr = c.opr
+            value = c.value
```
Comment (dpgaspar) [inline superset/daos/base.py]: nit(optional): col, opr, value = c.col, c.opr, c.value

File: superset/daos/base.py [hunk 19]
```diff
@@ -251,3 +455,205 @@ def filter_by(cls, **filter_by: Any) -> list[T]:
...
+        query = data_model.session.query(cls.model_cls)
+        query = cls._apply_base_filter(
+            query, skip_base_filter=skip_base_filter, data_model=data_model
+        )
+        if search and search_columns:
+            search_filters = []
+            for column_name in search_columns:
+                if hasattr(cls.model_cls, column_name):
+                    column = getattr(cls.model_cls, column_name)
+                    search_filters.append(cast(column, Text).ilike(f"%{search}%"))
```
[the remark this answers is not in the document]
Comment (dpgaspar) [reply superset/daos/base.py:582]: This is safe because we are using SQLAlchemy ORM
SQLAlchemy doesn't directly concatenate this into the SQL string. Instead, it:

Creates a parameterized query with placeholders
Passes the value '%' + search + '%' as a parameter separately from the SQL structure

The actual SQL sent to the database looks something like:

**Review by dpgaspar — APPROVED** [review]

File: superset/daos/base.py (hunk 16 above)
Comment (aminghadersohi) [reply superset/daos/base.py]: I like the idea of failing if the types aren't consistent but worried about the risk of these calls sites expecting it to work with mixed ids. 
  1. API Endpoints (Most likely sources of mixed IDs):
    - /charts/api.py - bulk operations accepting arrays of IDs
    - /dashboards/api.py - bulk operations accepting arrays of IDs
    - /tags/api.py - bulk operations accepting arrays of IDs
    - /row_level_security/api.py - bulk operations
  2. Command Classes (Delete operations):
    - All delete commands that accept lists of IDs from API endpoints
    - Examples: ChartDeleteCommand, DashboardDeleteCommand, etc.
  3. Tag Operations:
    - TagDAO.find_by_ids() when retrieving tagged objects
    - Could mix integer and UUID IDs if different object types use
  different ID schemes

If you think i shouldn't worry about that and we would prefer fail than warning i can add this: 

```
  if len(types_found) > 1:
      error_msg = f"Mixed ID types detected for {cls.model_cls.__name__ if 
  cls.model_cls else 'Unknown'}: {[t.__name__ for t in types_found]}"
      logger.error(error_msg)
      raise ValueError(error_msg)
```
